File: java_brain/src/main/java/org/netbrains/javabrain/neural/OutputNeuron.java
package org.netbrains.javabrain.neural;

import org.netbrains.javabrain.activation.Activation;

public class OutputNeuron extends WorkingNeuron {
    public OutputNeuron(int layerIndex, int index, Activation activation) {
        super(layerIndex, index, activation);
    }

    public void calculateDelta(double expectedValue) {
        double derivative = activation.derivative(netInput);
        double difference = expectedValue - value;
        delta = derivative * difference;
    }
}

File: cpp_brains/outputneuron.cpp
#include "outputneuron.h"

#ifdef TRAINING_DEBUG
#include <iostream>
#endif
#ifdef DEBUG
#include <iostream>
#endif

OutputNeuron::OutputNeuron(int layerIndex, int index, Activation *activation)
    :WorkingNeuron(layerIndex, index, activation)
{

}

void OutputNeuron::calculateDelta(double expected)
{
    double difference = expected - m_value;
    double derivative = m_activation->derivative(m_netInput);
    m_delta = derivative * difference;
#ifdef TRAINING_DEBUG
    std::cout << "delta for " << m_layerIndex << " " << m_index << " dif:" << difference << " D:" << m_delta << std::endl;
#endif
}
